### ISSUE TITLE
Pageimage, PagefileExtra, WebP does not recognize changes of its JPEG/PNG variation

### ISSUE THREAD

**Issue by horst-n** [processwire/processwire-issues#1497]
### Short description of the issue

When a pageimage once was called via $image->webp(), a webp-variation is created.
When the JPEG or PNG variation changes somehow BUT produces the SAME variation filename, the webp gets not updated.

### Expected behavior

The webp should show the exact same (updated) image like the current JPEG or PNG.

### Actual behavior

It keeps the old image and does not get updated.

### Optional: Suggestion for a possible fix

The following code is a new version of function webp() in Pageimage.php:
```
	/**
	 * Get WebP "extra" version of this Pageimage
	 *
	 * @return PagefileExtra
	 * @since 3.0.132
	 *
	 */
	public function webp() {
		$webp = $this->extras('webp');
		if(!$webp) {
			$webp = new PagefileExtra($this, 'webp');
			$webp->setArray($this->wire('config')->webpOptions);
			$this->extras('webp', $webp);
			$webp->addHookAfter('create', $this, 'hookWebpCreate'); 
		}

		// recognize changes of JPEG / PNG variation via timestamp comparison
		if($webp && 
			is_readable($webp->filename()) &&
			filemtime($webp->filename()) < filemtime($this->filename)
		) {
			$webp->create();
		}

		return $webp;
	}
```

### Steps to reproduce the issue

This is a template code to reproduce the behavior. Please compare the timestamps of the JPEG / PNG variation with its WebP variation in the second iteration, after the JPEG / PNG gets deleted & recreated:

```
<?php namespace ProcessWire;


ob_start();
?>
<style>
    table tr td {
        vertical-align: top;
        border-top: 1px solid #999;
    }
</style>

    <table style='width:100%;'>
        <tr>
            <td>timestamp jpeg<br/>timestamp webp</td>
            <td>name jpeg<br/>name webp</td>
            <td>&nbsp;</td>
        </tr>

<?php

    $config->webpOptions = array_merge($config->webpOptions, ['useSrcExt' => true]);
    $options = ['useSrcExt' => true];
    $image = $page->images->first();

        $img = $image->width(100, $options);

        $filenameJPG = $img->filename;
        $filenameWEBP = $img->webp()->filename;

?>
        <tr>
            <td><?php if(is_readable($filenameJPG)) echo date('d.m.Y H:i:s', filemtime($filenameJPG)) ?><br/><?php if(is_readable($filenameWEBP)) echo date('d.m.Y H:i:s', filemtime($filenameWEBP)) ?></td>
            <td><?php if(is_readable($filenameJPG)) echo basename($filenameJPG) ?><br/><?php if(is_readable($filenameWEBP)) echo basename($filenameWEBP) ?></td>
            <td><img src='<?=$img->url?>' alt='jpeg' /><br /><img src='<?=$img->url?>.webp' alt='webp' /></td>
        </tr>

<?php

        // NOW SLEEP FOR 2 SECONDS, DELETE ONLY THE JPEG AND RECALL THE PROCEDURE
        sleep(2);
        unlink($filenameJPG);

        // this should recreate the missing JPEG AND ALSO update the obsolete WEBP (see timestamps!)
        $img = $image->width(100, $options);
        $filenameWEBP = $img->webp()->filename;

?>
        <tr>
            <td><?php if(is_readable($filenameJPG)) echo date('d.m.Y H:i:s', filemtime($filenameJPG)) ?><br/><?php if(is_readable($filenameWEBP)) echo date('d.m.Y H:i:s', filemtime($filenameWEBP)) ?></td>
            <td><?php if(is_readable($filenameJPG)) echo basename($filenameJPG) ?><br/><?php if(is_readable($filenameWEBP)) echo basename($filenameWEBP) ?></td>
            <td><img src='<?=$img->url?>' alt='jpeg' /><br /><img src='<?=$img->url?>.webp' alt='webp' /></td>
        </tr>

    </table>
<?php

$content = ob_get_clean();

// echo $content;
```

### Setup/Environment

This behavior seems to exist with versions since 3.0.132, at least I have tested it yet with 3.0.175 and 3.0.190

**Comment by ryancramerdesign** [comment]
@horst-n Thanks for the report. Your test script does a manual unlink() call, and you'd have to use PW's API methods to remove the file before PW would identify it and re-create variations, etc. I'm not sure if that's the issue here but just wanted to mention it. I do think I can work with your example though and adjust it as needed to see if I can reproduce the issue. I can already see it looks like there may be a case where getVariations() is missing the .webp version, so that may be where the issue is since this method is also used to identify files to delete. Though if using a manual unlink() then it'll definitely miss them. While we could account that for checking filemtime, I think that may be more overhead than we'd want for something that would have to run on every usage, so I'd rather have the API remove/delete methods take care of this. I'll let you know what I find here.

**Comment by horst-n** [comment]
@ryancramerdesign Hi Ryan, I took this route (manually unlink the jpeg variation) as shorter or simpler to reproduce.
There are more then one posts in the support boards that describe the issue of not recreated webp variations when a jpeg/png variation has changed. So you can test with an interactively changed cropping variation too, if it only change the visual output but not the variation name. 
The same issue occurs (also) if the jpeg/png variation changed by the pw system. I just thought a newer timestamp would be enough to identify a point where it lacks the recognition for a webp recreation.
Posts from the forum that subscribe this issue with real pw system usage are:
This one says 'never deleted' but mainly means, and also one times says "get not refreshed"(!): 
  https://processwire.com/talk/topic/25524-webp-image-variations-never-deleted/

This one uses a ->width(450)->webp->url (simple width resizing, followed by a webp call). When the source of for the ->width(450) changes, the width(450) refreshes, but the following webp does not!
  https://processwire.com/talk/topic/23294-croppableimage-with-webp-support/page/2/#comment-219994
  https://processwire.com/talk/topic/23294-croppableimage-with-webp-support/?do=findComment&comment=220288

So there must be some point in the system where the jpeg variation gets not deleted but simply overwritten, or the jpeg variation gets deleted before recreation, but not its webp dependant.

**Comment by olafgleba** [comment]
I post this as a comment here as a addon to the topic issue. Hopefully it helps to isolate the issue.

### Summary
When a pageimage variation is created (either while editing in the PW backend or by calling $image->webp() in markup),- on deletion, the webp variants remains in the assets folder. They never gets deleted (besides the _never gets updated/recreated_ issue).

This occurs on a standalone core pageimage field just as with a core pageimage field within a RepeaterMatrix (v.5).

### Setup/Environment

- Tested with 3.0.148 and 3.0.192 Dev (localhost and different live environments)
- Image Sizer Engine: GD (2.1.1, 2.3.3) 
- Image Sizer Settings: default
- Webp Settings: default
- PHP 7.4.8
- Mysql 7.4.28
- BS/Browser: macos 10.15.7 (Catalina) / latest Firefox, Chrome, Safari

### Bare Examples

#### Backend creation, deletion

1. Set `$config->imageSizerOptions('webpAdd', true);` in site config.
2. Upload a image to a page (xxx.jpg). Two variants (Thumbs) are created (xxx.0x260.jpg, xxx.0x260.webp).
3. Mark and delete variant xxx.0x260.jpg in variation list. Leave variant overlay open.
4. xxx.0x260.jpg gets deleted from the assets folder **and** variant list immediately, xxx.0x260.webp gets deleted from the variant list immediately also, but actually **_remains in the assets folder_**.
5. With the variant overlay still open: After deletion, the filename of the remaining original file becomes the filename of the _deleted variant_ (e.g. xxx.0x260.jpg)...
6. Close the variation overlay and save the page. The ProcesswireLog shows a files-error `http://localhost/<path>/<to>/processwire/page/edit/?id=1&s=1&c=1    rename: Rename to pathname ($newName) that already exists: /<path>/<to>/site/assets/files/1/xxx.0x260.webp`. On second attempt (saving the page again), no error gets triggered no more. Opening the variant list again shows the correct filename of the original image.


**Note**: Nr. 5/6. are not necessary an issue (ajax/save state in conjunction with `display_errors = on`). But because it throws a message of type _Files-errors_, i just want to mention it, just in case...

#### Frontend creation, deletion:

1. Set `$config->imageSizerOptions('webpAdd', true);` in site config
2. Upload a image to a page (xxx.jpg).
3. Frontend: (Re)load web page to call creation (Markup s. below). Two more variants are created (xxx.480x640.jpg, xxx.480x640.webp).
4. Mark and delete variant xxx.480x640.jpg in variation list (Backend). Leave variant overlay open.
5. xxx.480x640.jpg gets deleted from the assets folder **and** variant list immediately, xxx.480x640.webp gets deleted from the variant list immediately also, but actually **_remains in the assets folder_**.
6. With variant overlay still open: After deletion, the filename of the remaining original becomes the  filename of the _deleted variant_ (e.g. xxx.480x640.jpg)...
7. Close the variation overlay and save the page. No error is logged. Opening the variant list again shows the correct filename of the original image.

**Note**: Nr. 6/7, same as above

#### Frontend creation, Markup

```
$page_image_palm = $page->image_head->size(480, 640);

<picture>
  <source srcset="<?= $page_image_palm->webp()->url; ?>" type="image/webp">
  <img src="<?= $page_image_palm->url; ?>" alt="">
</picture>
```

Kind of similar to [Markup used in Ryans post](https://processwire.com/blog/posts/webp-images-on-an-existing-site/)

**Comment by howdytom** [comment]
### Short description of the issue

ProcessWire does not remove generated webP files from the assets/files folder when deleting image variants. Related to PW Forum Topic: [webP variants are not automatically deleted](https://processwire.com/talk/topic/26041-webp-variants-are-not-automatically-deleted/#comment-223238)

### Expected behavior
Deleting the webP, JPG/PNG variations will delete all files within /assets/files/1134 folder.

### Actual behavior

Deleting the webP variations will remove them from the variants list, however all files still remain in the /assets/files/1134/ folder. This will increase dramatically disk space on bigger projects. 

### Steps to reproduce the issue

**Image field in a RepeaterMatrix v5:**

1. Within a Repeater image field → Variations → Delete checked → close the modal → Save

![1_PW-variations-list png 319a0f0152d2775bb2b6dd18ed4fbbe4](https://user-images.githubusercontent.com/9871520/164006456-6f0c71ce-e936-48e0-bb8a-eb93dae2ac03.png)

2. When I go to assets/files/1134/ all png variations are deleted. webP variations are still kept e.g. testgrafik-logodesign.1600x0se.webp, testgrafik-logodesign.1920x1080se.webp

![2_PW-asset-folder png 9163c9a81735349e34fc158a1a0222df](https://user-images.githubusercontent.com/9871520/164006529-80aafece-f53f-4bd6-892c-105644f67763.png)

3. When I open again the variations modal → the variation list is still empty, except the original file.

![3_PW-variations-list png 72dd81dd0bb1f60f28d263b565f65154](https://user-images.githubusercontent.com/9871520/164006674-bdbe0df6-7b02-45f9-a7e4-85fee215fe16.png)



*Note: Deleting the repeater Item, will also remove all files from assets/files directory!* 

### Setup/Environment

- ProcessWire 3.0.165 and ProcessWire dev 3.0.183
- PHP 7.3.29 and PHP 7.4.21
- GD 2.1.0

`$config->imageSizerOptions('webpAdd', false);`


RepeaterMatrix v5 Markup:

```
<?php if(count($page->img_test)) {
foreach($page->img_test as $image){
  $imgsettings = array(
    'quality' => 80,
    'cropping' => 'southeast'
  );
    $img_1920 = $image->size(1920, 1080, $imgsettings);
    $img_1600 = $image->size(1600, 0, $imgsettings);
    $img_500 = $image->size(500, 0, $imgsettings);
    echo "
    <picture>
      <source srcset='{$img_1920->webp->url}' loading='lazy' type='image/webp' alt='{$img_1920->description}' sizes='100vw' srcset='{$img_500->webp->url} 500w, {$img_1600->webp->url} 1600w, {$img_1920->webp->url} 1920w' class='canvas__fullsize-img'>
      <source srcset='{$img_1920->url}' loading='lazy' type='image/{$image->ext}' alt='{$img_1920->description}' sizes='100vw' srcset='{$img_500->url} 500w, {$img_1600->url} 1600w, {$img_1920->url} 1920w' class='canvas__fullsize-img'>
      <img src='{$img_1920->url}' loading='lazy' alt='{$img_1920->description}' sizes='100vw' srcset='{$img_500->url} 500w, {$img_1600->url} 1600w, {$img_1920->url} 1920w'>
    </picture>";
  }
}
```

Same issue occurs using core image fields with olafgleba's markup.

**Comment by matjazpotocnik** [comment]
Perhaps this is relevant? https://github.com/processwire/processwire-issues/issues/895

**Comment by ryancramerdesign** [comment]
@olafgleba I was able to duplicate and fix the issue you mentioned, thanks. @howdytom As far as I can tell, this is the same issue as the one reported by olafgleba, but just in a different context. I'm pretty sure the fix that was pushed will also resolve what you mentioned but let me know if you find otherwise.

**Comment by matjazpotocnik** [comment]
@horst-n Does Ryan's fix work for you? If it does, could you close this issue?

**Comment by olafgleba** [comment]
@ryancramerdesign @horst-n @howdytom @matjazpotocnik 
For me the fix pushed to PW Dev 30.199 solves the issue. Great...

**Comment by horst-n** [comment]
@matjazpotocnik Thanks for the reminder. The fix works perfect!
